gwcs/coordinate_frames.py

```diff
@@ -1,7 +1,107 @@
 # Licensed under a 3-clause BSD style license - see LICENSE.rst
 """
-Defines coordinate frames and ties them to data axes.
+This module defines coordinate frames for describing the inputs and/or outputs of a transform.
+
+In the following example, we have a two stage transform, with an input frame, an
+output frame and an intermediate frame.
+
+.. code-block::
+
+    ┌───────────────┐
+    │               │
+    │     Input     │
+    │     Frame     │
+    │               │
+    └───────┬───────┘
+            │
+      ┌─────▼─────┐
+      │ Transform │
+      └─────┬─────┘
+            │
+    ┌───────▼───────┐
+    │               │
+    │  Intermediate │
+    │     Frame     │
+    │               │
+    └───────┬───────┘
+            │
+      ┌─────▼─────┐
+      │ Transform │
+      └─────┬─────┘
+            │
+    ┌───────▼───────┐
+    │               │
+    │    Output     │
+    │     Frame     │
+    │               │
+    └───────────────┘
+
+
+Each frame instance is both metadata for the inputs/outputs of a transform and
+also a converter between those inputs/outputs and richer coordinate
+representations of those inputs/ouputs.
+
+For example, an output frame of type `~astropy.coordinates.SpectralCoord`
+provides metadata to the `.WCS` object such as the ``axes_type`` being
+``"SPECTRAL"`` and the unit of the output etc.  The output frame also provides a
+converter of the numeric output of the transform to a
+`~astropy.coordinates.SpectralCoord` object, by combining this metadata with the
+numerical values.
+
+``axes_order`` and conversion between objects and arguments
```

Comment:
This whole section of the documentation has been written to see if I actually understand the goals of `axes_order` and how it is intended to work. I am reasonably confident at this point that #455 is the result of things not working exactly as described here.

Can someone check that this is the desired end state before I go through and make it like this?

Reply:
That's my understanding too.